queryable-spec.ts

```diff
@@ -0,0 +1,141 @@
+
+import RDF from '@rdfjs/types';
+import {Algebra} from 'sparqlalgebrajs';
+import {EventEmitter} from 'events'; // TODO: refer to underlying interface, not to the class
+
+/*
+ * Helper union type 
+ */
+type termName = 'subject' | 'predicate' | 'object' | 'graph';
+
+/*
+ * Custom typings for the RDF/JS Stream interface as the current
+ * typings restrict the generic param Q to extensions of "BaseQuad",
+ * meaning it cannot be used for Bindings.
+ */
+export interface Stream<Q> extends EventEmitter {
+  read(): Q | null;
+}
+
+/*
+ * Map-like representation of Bindings as using plain objects could lead
+ * to collisions between variable names and object properties.
+ * Long-term goal: maintain compatibility with the native Map class.
+ */ 
+interface Bindings {
+  type: 'bindings';
+  get(variable: RDF.Variable): RDF.Term;
+  keys(): RDF.Variable[];
+  entries(): [RDF.Variable, RDF.Term][];
+  size: number;
+}
+
+/*
+ * Bindings objects are created using a dedicated factory, keeping in line
+ * with DataFactory.quad(). This also helps with facilitating support for
+ * immutability. Basic helper methods must also be provided for the most 
+ * common manipulations of bindings objects.
+ */
+interface BindingsFactory {
+  bindings(entries: [RDF.Variable, RDF.Term][]): Bindings;
+  merge(bindings: Bindings[]): Bindings;
+}
+
+/*
+ * Base interfaces to represent query results. These interfaces are generic
+ * with respect to the types of query metadata objects. These can be extended
+ * by implementors.
+ */
+
+interface BaseQueryResultMetadata<OrderItemsType> {
+  cardinality?: number; // Cardinality estimate
+  order?: OrderItemsType[];
+}
+
+interface BaseQueryResult<MetadataType extends BaseQueryResultMetadata<RDF.Variable | termName>> {
+  type: 'bindings' | 'quads' | 'boolean';
+  metadata(opts: Partial<Record<keyof MetadataType, boolean>>): Promise<MetadataType>;
+}
+
+interface QueryResultBindings<MetadataType extends BaseQueryResultMetadata<RDF.Variable>> extends BaseQueryResult<MetadataType> {
+  type: 'bindings';
+  bindings(): Promise<Bindings[]>;
+  stream(): Stream<Bindings>;
```

Comment:
We can also just make the stream lazy:
```suggestion
  stream: Stream<Bindings>;
```
(this is working out fine in practise in Comunica)

Reply:
I was thinking of going exactly the other way and make this into `stream(): Promise<Stream<Bindings>>`. I'm always wary of ending up with concatenated pass-through stream-like structures that end up impacting performance for no reason other than to have one less `await`. I don't know where I stand on this, curious to see what everybody else thinks.

Reply:
If an engine does _not_ use a promise internally you have the opposite problem: it's forced into the next tick even though its native stream is synchronously available.

Personally I would always prefer the least indirection possible in the interface. The function call helps readability because the code makes a choice whether to stream or just fetch metadata. So my preference is the original.

Reply:
This has been open for a while. Should we each rank these in order of preference? For me:

1. `stream(): Promise<Stream<Bindings>>`
2. `stream(): Stream<Bindings>`
3. `stream: Stream<Bindings>`

Reply:
No activity here, I'll go with `stream(): Promise<Stream<Bindings>>`. Happy to revisit and discuss further, just trying to keep the conversation going.